### PR TITLE
Issue 64  profile

### PR DESCRIPTION
**Pull request by adler-j**
I've done some of this, issues #64 and #63. Tests now run more than twice as fast.

* We had a few leftover `super()` in ntuples. Since we create a new space anytime we use `__getitem__`, anything where it is used can be considered "performance critical" and should use the named ancestor. Shaved off roughly 30% run-time on the tests by removing this.
* Roughly 20% of the test time is in `all_almost_equal`. I did a few optimizations here that should help (reduced from ~50%).
* In the vector classes we do `isinstance` checks on the input. I propose that we document that the vector classes are not intended to be used separately and that users should use `element`. That way we can remove a few `isinstance` calls here (that will always be true).
* There seems to be a `__instancecheck__` called for every ABCMeta inheriting class on construction. This uses about 20% of all runtime. There could also be a overhead in our `isinstance` calls since they seem to use a un-optimized python version instead of C code. We may actually want to consider totally dropping ABC and simply going for `raise NotImplementedError` in the methods.
* We used to spend about 20% of all run-time doing `int(indices)` as in:

    ```python
    try:
        return self.data[int(indices)]
    except TypeError:
        ...
    ```
    
    Since `int` is very expensive (if we want to use `future` ints), it is faster to do
    
    ```python
    if isinstance(indices, Integral):
        return self.data[indices]
    else:
        ...
    ```

### CODE REVIEW

**Comment by adler-j** [comment]
Added some further commits, not much more to add than the previous message. Running the test-suite now takes 2.28 seconds, compared to 5.38 on master. Shaved of 58% of the time.

Biggest culprits is still `isinstance` with abstract classes, and I would also say typechecking in the vector classes (which should be done away with since they shouldn't be used that way.

File: test/space/ntuples_test.py [hunk 1]
```diff
@@ -80,35 +80,26 @@ def _pos_array(fn):
...
-    coo_c = np.random.randint(0, fn.size, size=nnz)
+def _dense_matrix(fn):
+    """Create a dense positive definite Hermitian matrix for `fn`."""
+
     if np.issubdtype(fn.dtype, np.floating):
-        values = np.random.rand(nnz).astype(fn.dtype)
+        mat = np.random.rand(fn.size, fn.size).astype(fn.dtype)
     elif np.issubdtype(fn.dtype, np.integer):
-        values = np.random.randint(0, 10, nnz).astype(fn.dtype)
+        mat = np.random.randint(0, 10, (fn.size, fn.size)).astype(fn.dtype)
```
Comment (kohr-h) [inline test/space/ntuples_test.py]: `astype` can take a `copy=False` option which optimizes the case of the data type returned by the `np.random` functions.

**Comment by kohr-h** [comment]
I give up insisting on `super()`, at least in the case of `__init__()` and as long as we use the backported Python 2 thing. The native Py3 `super` should be fast. `ABCmeta` can go, I won't miss it. Agree also on the element type classes.

File: test/space/ntuples_test.py [hunk 2]
```diff
@@ -218,7 +209,7 @@ def test_vector_class_init(fn):
     arr = _array(fn)
 
     FnVector(fn, arr)
-
+"""
```
Comment (kohr-h) [inline test/space/ntuples_test.py]: Is this test removed for a reason or just forgotten to be put back in?

**Comment by kohr-h** [comment]
I added `copy=False` to the `astype()` calls, cutting off another 10 % in the Ntuples tests. ABCMeta will be dropped in the operator branch. Merge?

**Comment by adler-j** [comment]
Fine with merging then.

File: test/space/ntuples_test.py (hunk 2 above)
Comment (adler-j) [inline test/space/ntuples_test.py]: Was removed when I removed the tests in FnVector, but when I re-added that it shoulda been added back in.

**Pull request merged**